### DISCUSSION TITLE
Pytest How to pass .yaml file as argument to pytest command line

### DISCUSSION THREAD

**Discussion by dinesh-11-05** [pytest-dev/pytest#11905] [Q&A]
I have multiple .yaml files each containing shell or python commands and I want to execute them separately based on the need through command line arguments in pytest execution .

test_code.yaml

`**.yaml file content :**

test_execute_impala_query:
- change_dir_path: /home/a/
  sql_result: ksh -x /home/impala_query.sh /sandbox/home/impala_export_query.sql  /sandbox/home/actual_result.csv output_delimiter=,

test_file_compare:
 - file_compare_check: python3 /sandbox/pytest_execution/test_compare_file.py  --file1 /sandbox/venv/smoke_test/actual_result_prsn.csv --file2 /sandbox/smoke_test/expected_result.csv 

Actual Pytest script

test_code_check.py

`import pytest
import os
import parametrize_from_file
import pandas as pd
import subprocess
import sys

@parametrize_from_file
def test_execute_impala_query(change_dir_path,sql_result):
    os.chdir(change_dir_path)
    result = subprocess.run([sql_result],shell=True,check=True,stdout=subprocess.PIPE, universal_newlines=True)
    assert result.returncode == 0

@parametrize_from_file
def test_file_compare(file_compare_check):
    result = subprocess.run([file_compare_check],shell=True,check=True,stdout=subprocess.PIPE, universal_newlines=True)
    assert result.returncode == 0`

Here, the above code perfectly works fine but that .yaml filename I need to pass it through command line

I have to execute the below way to get the output separately. The below way I execute the pytest scripts and get the test comes for different set of .yaml file as argument to pass and generate the output . Can you please suggestions/inputs that would be very helpful

`pytest test_code_check.py --test_file test_code.yaml  

pytest test_code_check.py --test_file test_code_2.yaml 

pytest test_code_check.py --test_file test_code_3.yaml  

pytest test_code_check.py --test_file test_code_4.yaml`

**Comment by portyu9** [comment]
Hi @dinesh-11-05 

Yes. Register a pytest command-line option and perform the file parametrization during collection.

Since the YAML filename is only known after parsing the command line, instead of using the `@parametrize_from_file` decorator directly, use `parametrize_from_file.load_parameters()` from `pytest_generate_tests`.

For example:

```python
# conftest.py
from pathlib import Path

import parametrize_from_file as pff


def pytest_addoption(parser):
    parser.addoption(
        "--test-file",
        action="store",
        required=True,
        help="YAML file containing the test parameters",
    )


def pytest_configure(config):
    config.addinivalue_line(
        "markers",
        "from_param_file: load parameters from --test-file",
    )


def pytest_generate_tests(metafunc):
    if metafunc.definition.get_closest_marker("from_param_file") is None:
        return

    path = Path(
        metafunc.config.getoption("--test-file")
    ).resolve()

    names, params = pff.load_parameters(
        path,
        key=metafunc.function.__name__,
    )

    metafunc.parametrize(names, params)
```

Then remove `@parametrize_from_file` from these tests and mark them instead:

```python
import pytest


@pytest.mark.from_param_file
def test_execute_impala_query(change_dir_path, sql_result):
    ...


@pytest.mark.from_param_file
def test_file_compare(file_compare_check):
    ...
```

You can now run:

```text
pytest test_code_check.py --test-file test_code.yaml
pytest test_code_check.py --test-file test_code_2.yaml
pytest test_code_check.py --test-file test_code_3.yaml
```

`load_parameters()` uses the test function name as the YAML key in the example above, matching the structure you already have.

Please mark this answer as accepted if it helped, thank you.
